File: cpp/netkit/Box.cpp
//
// Created by dantezhu on 14-7-26.
// Copyright (c) 2014 dantezhu. All rights reserved.
//

#include <sstream>
#include <string.h>
#if defined(_WIN32) || (defined(CC_TARGET_PLATFORM) && CC_TARGET_PLATFORM==CC_PLATFORM_WIN32)
#include <winsock2.h>
#else
#include <netinet/in.h>
#endif
#include "Box.h"

#define box_mcp(dst, value) \
    memcpy(dst, &(value), sizeof(value)); \
    dst += sizeof(value);

namespace netkit {

Box::Box() {
    this->magic = MAGIC;
    this->version = 0;
    this->flag = 0;
    this->_placeholder_packet_len = 0;
    this->cmd = 0;
    this->ret = 0;
    this->sn = 0;
    this->_unpack_done = false;
}

Box::~Box() {}

int Box::pack(std::string& str) {
    str.resize(this->packetLen());

    char* buf = (char*)str.c_str();

    int i_temp;
    short s_temp;

    i_temp = htonl(this->magic);
    box_mcp(buf, i_temp);

    s_temp = htons(this->version);
    box_mcp(buf, s_temp);

    s_temp = htons(this->flag);
    box_mcp(buf, s_temp);

    i_temp = htonl(this->packetLen());
    box_mcp(buf, i_temp);

    i_temp = htonl(this->cmd);
    box_mcp(buf, i_temp);

    i_temp = htonl(this->ret);
    box_mcp(buf, i_temp);

    i_temp = htonl(this->sn);
    box_mcp(buf, i_temp);

    memcpy(buf, this->getBody(), this->bodyLen());
    buf += this->bodyLen();

    return this->packetLen();
}

int Box::unpack(const char* buf, int length) {
    return this->_unpack(buf, length, true);
}


int Box::check(const char* buf, int length) {
    return this->_unpack(buf, length, false);
}

int Box::_unpack(const char* buf, int length, bool save) {
    if (length < this->headerLen()) {
        // 连包头都不够
        return 0;
    }

    int magic_ = ntohl(*((int*)buf));
    buf += sizeof(int);

    if (magic_ != MAGIC) {
        // 说明magic不对
        return -1;
    }

    short version_ = ntohs(*((short*)buf));
    buf += sizeof(short);

    short flag_ = ntohs(*((short*)buf));
    buf += sizeof(short);

    int packet_len_ = ntohl(*((int*)buf));
    buf += sizeof(int);

    int cmd_ = ntohl(*((int*)buf));
    buf += sizeof(int);

    int ret_ = ntohl(*((int*)buf));
    buf += sizeof(int);

    int sn_ = ntohl(*((int*)buf));
    buf += sizeof(int);

    if (packet_len_ > length) {
        // 还没收完，继续
        return 0;
    }

    if (!save) {
        return packet_len_;
    }

    this->magic = magic_;
    this->version = version_;
    this->flag = flag_;
    this->_placeholder_packet_len = packet_len_;
    this->cmd = cmd_;
    this->ret = ret_;
    this->sn = sn_;

    this->setBody(buf, packet_len_ - this->headerLen());

    this->_unpack_done = true;

    return packet_len_;
}

const char* Box::getBody() {
    return this->_body.c_str();
}

const std::string& Box::getStringBody() {
    return this->_body;
}

void Box::setBody(const char* p_body, int body_len) {
    if (!p_body) {
        this->_body.resize(0);
        return;
    }

    this->_body.assign(p_body, body_len);
}

void Box::setBody(const std::string& str) {
    setBody(str.c_str(), str.size());
}

int Box::headerLen() {
    return sizeof(this->magic) + sizeof(this->version) + sizeof(this->flag) + sizeof(this->_placeholder_packet_len) + \
        sizeof(this->cmd) + sizeof(this->ret) + sizeof(this->sn);
}

int Box::bodyLen() {
    return this->_body.size();
}

int Box::packetLen() {
    return this->headerLen() + this->bodyLen();
}

bool Box::unpackDone() {
    return this->_unpack_done;
}

std::string Box::toString() {
    std::stringstream ss;
    ss << "magic: " << this->magic;
    ss << ", version: " << this->version;
    ss << ", flag: " << this->flag;
    ss << ", cmd: " << this->cmd;
    ss << ", ret: " << this->ret;
    ss << ", sn: " << this->sn;
    ss << ", bodyLen: " << this->bodyLen();

    return ss.str();
}

}

File: java/src/cn/vimer/netkit/Box.java
package cn.vimer.netkit;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.ByteArrayOutputStream;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.ByteArrayInputStream;
import java.io.IOException;


// 与python的netkit不同
// java版只认网络字节序，因为java的所有存储都是以网络字节序
// 只支持packet_len

public class Box implements IBox {

    // 魔法数字
    public final static int MAGIC = 2037952207;

    // 默认就用这个值
    public int magic = MAGIC;
    public short version;
    public short flag;
    // 这个只用来在网络上传输，放在这里仅作占位符
    private int _placeholder_packet_len;
    public int cmd;
    public int ret;
    public int sn;

    // 包体
    public byte[] body = null;

    // 如果解析完毕会被设置为true
    private boolean _unpack_done = false;

    // 压包
    public byte[] pack() {

        ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream(this.packetLen());
        DataOutputStream outputStream = new DataOutputStream(byteArrayOutputStream);

        try{
            outputStream.writeInt(magic);
            outputStream.writeShort(version);
            outputStream.writeShort(flag);
            outputStream.writeInt(packetLen());
            outputStream.writeInt(cmd);
            outputStream.writeInt(ret);
            outputStream.writeInt(sn);
            if (body != null) {
                outputStream.write(body);
            }

            outputStream.flush();
        }
        catch (Exception e) {
            return null;
        }

        return byteArrayOutputStream.toByteArray();
    }

    // 解包
    public int unpack(byte[] buf) {
        return unpack(buf, 0, buf == null ? 0:buf.length);
    }

    public int unpack(byte[] buf, int offset, int length) {
        return _unpack(buf, offset, length, true);
    }

    //>0: 成功生成obj，返回了使用的长度，即剩余的部分buf要存起来
    //<0: 报错
    //0: 继续收
    public int _unpack(byte[] buf, int offset, int length, boolean save) {

        if (buf == null) {
            return -100;
        }

        length = buf.length > length ? length : buf.length;

        if (length < this.headerLen()) {
            return 0;
        }

        DataInputStream inputStream = new DataInputStream(new ByteArrayInputStream(buf, offset, length));

        int _magic;
        short _version;
        short _flag;
        int _packet_len;
        int _cmd;
        int _ret;
        int _sn;
        int _body_len;
        byte[] _body = null;

        try {
            _magic = inputStream.readInt();
            _version = inputStream.readShort();
            _flag = inputStream.readShort();
            _packet_len = inputStream.readInt();
            _cmd = inputStream.readInt();
            _ret = inputStream.readInt();
            _sn = inputStream.readInt();
        }
        catch (Exception e) {
            return -1;
        }

        if (_magic != MAGIC) {
            // 包错误
            return -2;
        }

        if (length < _packet_len) {
            // 继续收
            return 0;
        }

        _body_len = _packet_len - this.headerLen();

        try{
            if (_body_len > 0) {
                _body = new byte[_body_len];
                int len = inputStream.read(_body);

                // 说明解出来的body长度不对
                if (len != _body_len) {
                    System.out.println(len);
                    return -3;
                }
            }
        }
        catch (Exception e) {
            // 解body异常
            return -4;
        }

        if (!save) {
            return _packet_len;
        }

        this.magic = _magic;
        this.version = _version;
        this.flag = _flag;
        this._placeholder_packet_len = _packet_len;
        this.cmd = _cmd;
        this.ret = _ret;
        this.sn = _sn;
        this.body = _body;

        this._unpack_done = true;

        return _packet_len;
    }

    // 检查
    public int check(byte[] buf) {
        return check(buf, 0, buf == null ? 0:buf.length);
    }

    // 检查
    public int check(byte[] buf, int offset, int length) {
        return _unpack(buf, offset, length, false);
    }

    public int headerLen() {
        return 4 * 5 + 2 * 2;
    }

    public int bodyLen() {
        if (body != null) {
            return body.length;
        }

        return 0;
    }

    public int packetLen() {
        return headerLen() + bodyLen();
    }

    public boolean unpackDone() {
        return _unpack_done;
    }

    public JSONObject getJson() throws IOException, JSONException {
        if (body == null) {
            return null;
        }

        return new JSONObject(new String(body, "UTF-8"));
    }

    public void setJson(JSONObject jsonObject) {
        if (jsonObject == null) {
            body = null;
        }

        body = jsonObject.toString().getBytes();
    }

    public String toString() {
        String out = "";
        out += "magic: " + magic;
        out += ", version: " + version;
        out += ", flag: " + flag;
        out += ", packetLen: " + packetLen();
        out += ", cmd: " + cmd;
        out += ", ret: " + ret;
        out += ", sn: " + sn;
        out += ", bodyLen: " + bodyLen();

        return out;
    }
}
